benchmark/utils/runner.hpp

```diff
@@ -0,0 +1,209 @@
+                     std::shared_ptr<Timer> timer, State& state,
+                     const std::string& operation,
+                     json& operation_case) const = 0;
+
+    /** Post-process test case info. */
+    virtual void postprocess(json& test_case) const {}
+};
+
+
+template <typename State>
+void run_test_cases(const Benchmark<State>& benchmark,
+                    std::shared_ptr<gko::Executor> exec,
+                    std::shared_ptr<Timer> timer, json& test_cases)
+{
+    if (!test_cases.is_array()) {
+        if (benchmark.should_print()) {
+            std::cerr
+                << "Input has to be a JSON array of benchmark configurations:\n"
+                << benchmark.get_example_config() << std::endl;
+        }
+        std::exit(1);
+    }
+    for (const auto& test_case : test_cases) {
+        if (!test_case.is_object() || !benchmark.validate_config(test_case)) {
+            if (benchmark.should_print()) {
+                std::cerr << "Invalid test case:\n"
+                          << std::setw(4) << test_case << "\nInput format:\n"
+                          << benchmark.get_example_config() << std::endl;
+            }
+            std::exit(2);
+        }
+    }
+
+    auto profiler_hook = create_profiler_hook(exec, benchmark.should_print());
+    if (profiler_hook) {
+        exec->add_logger(profiler_hook);
+    }
+    auto annotate =
+        [profiler_hook](const char* name) -> gko::log::profiling_scope_guard {
+        if (profiler_hook) {
+            return profiler_hook->user_range(name);
+        }
+        return {};
+    };
+
+    for (auto& test_case : test_cases) {
+        try {
+            // set up benchmark
+            if (!test_case.contains(benchmark.get_name())) {
+                test_case[benchmark.get_name()] = json::object();
+            }
+            if (benchmark.should_print()) {
+                std::clog << "Running test case\n"
+                          << std::setw(4) << test_case << std::endl;
+            }
+            auto test_case_state = benchmark.setup(exec, test_case);
+            auto test_case_str = benchmark.describe_config(test_case);
+            auto test_case_range = annotate(test_case_str.c_str());
+            auto& benchmark_case = test_case[benchmark.get_name()];
+            for (const auto& operation_name : benchmark.get_operations()) {
+                if (benchmark_case.contains(operation_name) &&
+                    !FLAGS_overwrite) {
+                    continue;
+                }
+                benchmark_case[operation_name] = json::object();
+                if (benchmark.should_print()) {
+                    std::clog << "\tRunning " << benchmark.get_name() << ": "
+                              << operation_name << std::endl;
+                }
+                auto& operation_case = benchmark_case[operation_name];
+                try {
+                    auto operation_range = annotate(operation_name.c_str());
```

Comment:
I think it is still useful to have a single repetition annotated, since there are lot of other operations happening around it, but I omit warmup ranges if `FLAGS_warmup == 0`.

Reply:
I forget what it is for actually.
I think it destroyed the iteration control which reduces the sync point.
We do not expect less overhead if we enable the profiler/annotation.
It should be okay if we do not have these with `none` profiler.